_bleio/characteristic_buffer.py

```diff
@@ -0,0 +1,140 @@
+
+* Author(s): Dan Halbert for Adafruit Industries
+"""
+from __future__ import annotations
+from typing import Union
+
+import asyncio
+import queue
+import time
+
+import _bleio.adapter_ as adap
+from _bleio.characteristic import Characteristic
+
+Buf = Union[bytes, bytearray, memoryview]
+
+
+class CharacteristicBuffer:
+    """Accumulates a Characteristic's incoming values in a FIFO buffer."""
+
+    def __init__(
+        self,
+        characteristic: Characteristic,
+        *,
+        timeout: float = 1,
+        buffer_size: int = 64
+    ):
+
+        """Monitor the given Characteristic. Each time a new value is written to the Characteristic
+        add the newly-written bytes to a FIFO buffer.
+
+        :param Characteristic characteristic: The Characteristic to monitor.
+          It may be a local Characteristic provided by a Peripheral Service,
+          or a remote Characteristic in a remote Service that a Central has connected to.
+        :param int timeout:  the timeout in seconds to wait for the first character
+          and between subsequent characters.
+        :param int buffer_size: Size of ring buffer that stores incoming data coming from client.
+          Must be >= 1."""
+        self._characteristic = characteristic
+        self._timeout = timeout
+        self._queue = queue.Queue(buffer_size)
+        characteristic.add_notify_callback(self._notify_callback)
+
+    def _notify_callback(self, data: Buf) -> None:
+        # Add data bytes to queue, one at a time.
+        if self._queue.full():
+            # Discard oldest data to make room.
+            while self._queue.qsize() > len(data):
+                self._queue.get_nowait()
+
+        for b in data:
+            try:
+                self._queue.put_nowait(b)
+            except queue.Full:
+                return
+
+    def read(self, nbytes: int = None) -> Union[Buf, None]:
+        """Read characters.  If ``nbytes`` is specified then read at most that many
+        bytes. Otherwise, read everything that arrives until the connection
+        times out. Providing the number of bytes expected is highly recommended
+        because it will be faster.
+
+        :return: Data read
+        :rtype: bytes or None
+        """
+        b = bytearray()
+        end = time.time() + self._timeout
+        while len(b) < nbytes and time.time() < end:
+            try:
+                b.append(self._queue.get_nowait())
+            except queue.Empty:
+                # Let the BLE code run for a bit, and try again.
+                adap.adapter.await_bleak(asyncio.sleep(0.1))
+        return b
+
+    def readinto(self, buf: Buf) -> Union[int, None]:
+        """Read bytes into the ``buf``. Read at most ``len(buf)`` bytes.
+        :return: number of bytes read and stored into ``buf``
+        :rtype: int or None (on a non-blocking error)
+        """
+        bytes_read = self.read(len(buf))
+        buf[0 : len(bytes_read)] = bytes_read
```

Comment:
Please swap these and call readinto from read. Let's not always do a bytearray allocation and then a copy.

I know this is on full computers but I worry folks will pick up on this pattern and use it on smaller devices.